### PR TITLE
OWImpute: fix model based imputer. Fixes #1082.

### LINKED ISSUE #1082: Impute widget crashes on learner input
Impute widget crashes when provided with a learner input. Should automatically switch to 'Model-based imputer' and use it. Also the widget should check for model compatibility (classification or regression).

### PR DESCRIPTION
**Pull request by s-alexey**
Switch to Model-based imputer in case learner is provided.

Added learner adequacy checker.

### CODE REVIEW

File: Orange/preprocess/impute.py [hunk 1]
```diff
@@ -107,11 +107,14 @@ def __call__(self, data, variable):
...
-        return variable.copy(
-            compute_value=ReplaceUnknownsModel(variable, model))
 
+        if self.learner.check_learner_adequacy(domain):
+            model = self.learner(data)
+            assert model.domain.class_var == variable
+            return variable.copy(
+                compute_value=ReplaceUnknownsModel(variable, model))
+        else:
+            return variable.copy()
```
Comment (astaric) [inline Orange/preprocess/impute.py]: When learner on input is not compatible with the variable, the variable does not get imputed.

I would rather move the check to the imputation widget. When the learner on input does not support variable type, a Model imputer with default_learner could be constructed.

**Comment by astaric** [comment]
I like the fix as it is simple and straightforward. The problem is that the change on the Model imputer now allows the user to construct a model imputer for a variable of the wrong type, which would not impute anything. This is not desired.

I would rather move the check to the widget, and use some kind of default_learner for the variables with incompatible type (rather than not impute them).

If Model based imputer option was renamed to Model based imputer (<learner>), this would also signal to the user which learner will be used for imputation. When a given learner only supports one kind of variables, text could dynamically change to "Model base imputer (continuous: <learner1>, discrete: <learner2>" or something like this.

**Comment by s-alexey** [comment]
@astaric I have changed my PR. 
It's not quite what you suggested but hope you will like it.

File: Orange/preprocess/impute.py [hunk 2]
```diff
@@ -99,18 +145,46 @@ def __call__(self, data):
...
+
+    def str(self, var):
+        return super().str(var) + " ({})".format(self.learner)
+
+    def copy(self):
+        return Model(self.learner)
+
+    @property
+    def support_discrete(self):
```
Comment (astaric) [inline Orange/preprocess/impute.py]: What about the classes that support both (simple tree?)

File: Orange/preprocess/impute.py [hunk 3]
```diff
@@ -99,18 +145,46 @@ def __call__(self, data):
...
-        return variable.copy(
-            compute_value=ReplaceUnknownsModel(variable, model))
+
+        if self.learner.check_learner_adequacy(domain):
+            model = self.learner(data)
+            assert model.domain.class_var == variable
+            return variable.copy(
+                compute_value=ReplaceUnknownsModel(variable, model))
+        else:
+            return variable.copy()
```
Comment (astaric) [inline Orange/preprocess/impute.py]: I still do not like this. If I call an instance of a Model imputer with data, I would like to have imputed data. If the learner does not support variable type and variable contains missing values, this should IMO raise an exception.

Or maybe model could accept multiple learners as a parameter, one for continuous and one for discrete?

**Comment by astaric** [comment]
I generally like the new hierarchy, although I cannot decide whether it should have and "abstract" class on top instead of the DontImpute imputer.

BTW, have you considered having a kind of a registry (https://github.com/biolab/orange3/blob/master/Orange/util.py#L105) to collect all available imputers (instead of hardcoding them in the widget)?

File: Orange/widgets/data/owimpute.py [hunk 4]
```diff
@@ -220,192 +104,163 @@ def __init__(self):
...
+
         self.unconditional_commit()
 
     def set_learner(self, learner):
         self.learner = learner
+        self.METHODS[self.MODEL_BASED_IMPUTER].learner = learner
+
+        button = self.default_button_group.button(self.MODEL_BASED_IMPUTER)
+        variable_button = self.variable_button_group.button(self.MODEL_BASED_IMPUTER)
+        if self.learner is None:
```
Comment (astaric) [inline Orange/widgets/data/owimpute.py]: Currently, Model based imputer uses a default learner (simple tree), and the input is only used as a way to override the learner.

File: Orange/preprocess/impute.py (hunk 3 above)
Comment (s-alexey) [inline Orange/preprocess/impute.py]: I don't like too, I have been meaning to replace it with an exception.

File: Orange/widgets/data/owimpute.py (hunk 4 above)
Comment (s-alexey) [inline Orange/widgets/data/owimpute.py]: I can add default learner or add a new `SimpleTree` impute method.

File: Orange/preprocess/impute.py (hunk 2 above)
Comment (s-alexey) [inline Orange/preprocess/impute.py]: They are supported. 
It's the second time I have faced this problem (check learner type). Is there a good way to do it?

Comment (astaric) [inline Orange/preprocess/impute.py]: What about constructing a dummy domain and using check_learner_adequacy?

File: Orange/widgets/data/owimpute.py (hunk 4 above)
Comment (astaric) [inline Orange/widgets/data/owimpute.py]: I would prefer a default learner. This way, everything "just works", and if user wants to change the learner, he can.

File: Orange/widgets/data/owimpute.py [hunk 5]
```diff
@@ -220,192 +113,169 @@ def __init__(self):
...
     def set_data(self, data):
         self.closeContext()
-        self.clear()
+        self.varmodel[:] = []
+        self.variable_methods = {}
+        self.modified = False
         self.data = data
+
         if data is not None:
             self.varmodel[:] = data.domain.variables
```
Comment (s-alexey) [inline Orange/widgets/data/owimpute.py]: Is there a good way to filter variables without missed values?

File: Orange/preprocess/impute.py [hunk 6]
```diff
@@ -26,7 +27,61 @@ def transform(self, c):
...
 
-class Average:
+class BaseImputeMethod:
+    name = ""
+    short_name = ""
+    description = ""
+    format = "{var.name} -> {short_name}"
+    columns_only = False
+
+    def __init__(self, *args, **kwargs):
```
Comment (astaric) [inline Orange/preprocess/impute.py]: If __init__ does nothing, you can ommit it.

File: Orange/preprocess/impute.py [hunk 7]
```diff
@@ -26,7 +27,61 @@ def transform(self, c):
...
+        pass
+
+    @abstract
+    def __call__(self, data, variable):
+        raise NotImplementedError()
```
Comment (astaric) [inline Orange/preprocess/impute.py]: @abstract decorator will raise NotImplementedError. The body of the method could include a helpful docstring instead (describing what is the purpose of the method and its parameters).

File: Orange/preprocess/impute.py [hunk 8]
```diff
@@ -26,7 +27,61 @@ def transform(self, c):
...
+
+    def str(self, var):
+        return self.format.format(var=var, short_name=self.short_name)
```
Comment (astaric) [inline Orange/preprocess/impute.py]: This method could use a docstring describing what it is supposed to return and how does that differ from __str__. Depending on the purpose, a more descriptive method name could be used.

File: Orange/preprocess/impute.py [hunk 9]
```diff
@@ -26,7 +27,61 @@ def transform(self, c):
...
+
+    def __str__(self):
+        return self.name
+
+    def copy(self):
+        return self
+
+    def __eq__(self, other):
+        return type(self) == type(other)
```
Comment (astaric) [inline Orange/preprocess/impute.py]: Could use a docstring/comment on why is this a good idea.

File: Orange/preprocess/impute.py [hunk 10]
```diff
@@ -152,7 +240,13 @@ def transform(self, column):
         return numpy.where(mask, unknown, values)
 
 
-class AsValue:
+class AsValue(BaseImputeMethod):
+    # name = "Value"
```
Comment (astaric) [inline Orange/preprocess/impute.py]: This comments can be removed.

File: Orange/widgets/data/owimpute.py [hunk 11]
```diff
@@ -1,169 +1,53 @@
...
-def radio_button(text="", checked=False, group=None, group_id=None):
-    button = QRadioButton(text, checked=checked)
-    if group is not None:
-        group.addButton(button, )
-        if group_id is not None:
-            group.setId(button, group_id)
-    return button
+    method = impute.DoNotImpute()
 
+    # def str(self, var):
```
Comment (astaric) [inline Orange/widgets/data/owimpute.py]: Commented lines can be removed.

File: Orange/widgets/data/owimpute.py (hunk 5 above)
Comment (astaric) [inline Orange/widgets/data/owimpute.py]: Not that I know of.

**Comment by astaric** [comment]
The titles for sections Default method and Individual attribute settings are not displayed correctly on OSX.

screen shot 2016-05-09 at 09 33 28

File: Orange/widgets/data/owimpute.py [hunk 12]
```diff
@@ -220,192 +113,169 @@ def __init__(self):
...
-        else:
-            data = None
+
+            self.warning(1)
+            with self.progressBar(len(self.varmodel)) as progress:
+                for i, var in enumerate(self.varmodel):
+                    method = self.variable_methods.get(i, self.default_method)
+
+                    if not method.check_supports_variable(var):
+                        self.warning(1, "Default method has ignored some variables.")
```
Comment (astaric) [inline Orange/widgets/data/owimpute.py]: Default method could not impute some of the variables?

**Comment by astaric** [comment]
I have made some comments in the code. The only "major" issue I have found is the layout issue above. The rests seems ready for merge to me.

**Comment by s-alexey** [comment]
@astaric I have made some changes, can you review it and check the layout issue?

**Comment by astaric** [comment]
The layout issue remains. @kernc, can you check if it looks ok on your computer?

**Comment by kernc** [comment]
@astaric The layout seems ok here.
screenshot - 05102016 - 01 43 04 pm

File: Orange/preprocess/impute.py [hunk 13]
```diff
@@ -26,7 +28,65 @@ def transform(self, c):
         return numpy.where(numpy.isnan(c), self.value, c)
 
 
-class Average:
+class BaseImputeMethod:
+    name = ""
+    short_name = ""
+    description = ""
+    format = "{var.name} -> {short_name}"
```
Comment (kernc) [inline Orange/preprocess/impute.py]: How about specifying this as `"{var.name} → {self.short_name}"`. Then `format_variable()` method below could call format it with `self.format.format(var=var, self=self)` and there would be no reason to override the method if `format` is already overriden?

File: Orange/preprocess/impute.py [hunk 14]
```diff
@@ -26,7 +28,65 @@ def transform(self, c):
...
+        return self.format.format(var=var, short_name=self.short_name)
+
+    def __str__(self):
+        return self.name
+
+    def copy(self):
+        return self
+
+    @classmethod
+    def check_supports_variable(cls, variable):
```
Comment (kernc) [inline Orange/preprocess/impute.py]: `supports_variable()` is shorter and grammatically clearer; the method either supports variable or not supports variable.

File: Orange/preprocess/impute.py [hunk 15]
```diff
@@ -26,7 +28,65 @@ def transform(self, c):
...
+
+    def __call__(self, data, variable):
+        index = data.domain.index(variable)
+        return numpy.isnan(data[:, index]).reshape(-1)
+
+
+class Average(BaseImputeMethod):
+    name = "Average/Most frequent"
+    short_name = "average"
+    description = "Replace with average/modus for the column"
```
Comment (kernc) [inline Orange/preprocess/impute.py]: Modus is [mode](https://en.wikipedia.org/wiki/Mode_%28statistics%29) in English.

File: Orange/widgets/data/owimpute.py [hunk 16]
```diff
@@ -176,38 +57,47 @@ class OWImpute(OWWidget):
...
+        box = QtGui.QGroupBox(title=self.tr("Default method"), flat=False)
+        main_layout.addWidget(box)
+
+        button_group = QtGui.QButtonGroup()
+        for i, method in enumerate(self.METHODS):
+            if not method.columns_only:
+                button = QtGui.QRadioButton(method.name)
+                button.setChecked(i == self.default_method_index)
+                button_group.addButton(button, i)
+                main_layout.layout().addWidget(button)
```
Comment (kernc) [inline Orange/widgets/data/owimpute.py]: `box.layout().addWidget(button)`? That's what was before.

File: Orange/widgets/data/owimpute.py [hunk 17]
```diff
@@ -220,192 +110,170 @@ def __init__(self):
...
-        self.value_stack = value_stack = QStackedLayout()
-        self.value_combo = QComboBox(
+        self.value_combo = QtGui.QComboBox(
             minimumContentsLength=8,
-            sizeAdjustPolicy=QComboBox.AdjustToMinimumContentsLength,
+            sizeAdjustPolicy=QtGui.QComboBox.AdjustToMinimumContentsLength,
             activated=self._on_value_changed)
-        self.value_line = QLineEdit(editingFinished=self._on_value_changed)
-        self.value_line.setValidator(QDoubleValidator())
+        self.value_double = QtGui.QDoubleSpinBox(
```
Comment (kernc) [inline Orange/widgets/data/owimpute.py]: Does this value get saved as a Setting?

File: Orange/widgets/data/owimpute.py [hunk 18]
```diff
@@ -220,192 +110,170 @@ def __init__(self):
...
+            button = QtGui.QRadioButton(text=method.name)
+            button_group.addButton(button, i)
+            method_layout.addWidget(button)
 
-        self.value_stack = value_stack = QStackedLayout()
-        self.value_combo = QComboBox(
+        self.value_combo = QtGui.QComboBox(
             minimumContentsLength=8,
-            sizeAdjustPolicy=QComboBox.AdjustToMinimumContentsLength,
+            sizeAdjustPolicy=QtGui.QComboBox.AdjustToMinimumContentsLength,
```
Comment (kernc) [inline Orange/widgets/data/owimpute.py]: Forward compatible, it's better to import and refer to `QObjects` directly. In Qt 5, for example, `QComboBox` is in `QtWidgets` namespace (module).

File: Orange/widgets/data/owimpute.py [hunk 19]
```diff
@@ -220,192 +110,170 @@ def __init__(self):
...
-        assert self.METHODS[-1].short == "value"
+        button_group = QtGui.QButtonGroup()
+        for i, method in enumerate(self.METHODS):
+            button = QtGui.QRadioButton(text=method.name)
+            button_group.addButton(button, i)
+            method_layout.addWidget(button)
 
-        self.value_stack = value_stack = QStackedLayout()
-        self.value_combo = QComboBox(
+        self.value_combo = QtGui.QComboBox(
```
Comment (kernc) [inline Orange/widgets/data/owimpute.py]: The selected value here should be saved as a Setting as well.

**Comment by astaric** [comment]
The box surrounding default methods/individual attribute settings is now (again) shown on OSX, so the layout issue is gone.

File: Orange/preprocess/impute.py [hunk 20]
```diff
@@ -26,7 +28,65 @@ def transform(self, c):
...
+
+    def __call__(self, data, variable):
+        index = data.domain.index(variable)
+        return numpy.isnan(data[:, index]).reshape(-1)
+
+
+class Average(BaseImputeMethod):
+    name = "Average/Most frequent"
+    short_name = "average"
+    description = "Replace with average/mode for the column"
```
Comment (kernc) [inline Orange/preprocess/impute.py]: average value / mode _of_

**Comment by astaric** [comment]
If @kernc has no more comments, I am going to merge this today. If you were planning on changing anything, please let me know.

File: Orange/widgets/data/owimpute.py (hunk 19 above)
Comment (s-alexey) [inline Orange/widgets/data/owimpute.py]: It wasn't saved as a Settings previously and I'm not sure we should save it now. It's an index and actual value depends on selected variable.

**Comment by s-alexey** [comment]
I'm not planning any changes.

**Pull request merged**